File: services/product-service/src/main/java/com/ecommerce/product/model/Product.java
package com.ecommerce.product.model;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.index.TextIndexed;

import jakarta.validation.constraints.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

@Document(collection = "products")
public class Product {
    
    @Id
    private String id;
    
    @NotBlank(message = "Product name is required")
    @Size(min = 1, max = 200, message = "Product name must be between 1 and 200 characters")
    @TextIndexed(weight = 2)
    private String name;
    
    @NotBlank(message = "Product description is required")
    @Size(min = 10, max = 2000, message = "Description must be between 10 and 2000 characters")
    @TextIndexed
    private String description;
    
    @NotBlank(message = "SKU is required")
    @Indexed(unique = true)
    private String sku;
    
    @NotNull(message = "Price is required")
    @DecimalMin(value = "0.0", inclusive = false, message = "Price must be greater than 0")
    @Digits(integer = 10, fraction = 2, message = "Price format is invalid")
    private BigDecimal price;
    
    @DecimalMin(value = "0.0", message = "Compare price must be greater than or equal to 0")
    @Digits(integer = 10, fraction = 2, message = "Compare price format is invalid")
    private BigDecimal comparePrice;
    
    @NotNull(message = "Category ID is required")
    @Indexed
    private String categoryId;
    
    @NotBlank(message = "Brand is required")
    @Indexed
    private String brand;
    
    @NotEmpty(message = "At least one image is required")
    private List<String> images;
    
    private List<String> tags;
    
    @NotNull(message = "Weight is required")
    @DecimalMin(value = "0.0", inclusive = false, message = "Weight must be greater than 0")
    private BigDecimal weight;
    
    private Dimensions dimensions;
    
    private Map<String, Object> attributes; // Color, Size, Material, etc.
    
    private List<Variant> variants;
    
    @Min(value = 0, message = "Stock quantity cannot be negative")
    private Integer stockQuantity = 0;
    
    @Min(value = 0, message = "Low stock threshold cannot be negative")
    private Integer lowStockThreshold = 5;
    
    private Boolean trackInventory = true;
    
    private Boolean isActive = true;
    
    private Boolean isFeatured = false;
    
    @DecimalMin(value = "0.0", message = "Rating cannot be negative")
    @DecimalMax(value = "5.0", message = "Rating cannot be greater than 5")
    private BigDecimal averageRating = BigDecimal.ZERO;
    
    @Min(value = 0, message = "Review count cannot be negative")
    private Integer reviewCount = 0;
    
    @CreatedDate
    private LocalDateTime createdAt;
    
    @LastModifiedDate
    private LocalDateTime updatedAt;
    
    // Constructors
    public Product() {}
    
    public Product(String name, String description, String sku, BigDecimal price, 
                   String categoryId, String brand, List<String> images) {
        this.name = name;
        this.description = description;
        this.sku = sku;
        this.price = price;
        this.categoryId = categoryId;
        this.brand = brand;
        this.images = images;
    }
    
    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public String getSku() { return sku; }
    public void setSku(String sku) { this.sku = sku; }
    
    public BigDecimal getPrice() { return price; }
    public void setPrice(BigDecimal price) { this.price = price; }
    
    public BigDecimal getComparePrice() { return comparePrice; }
    public void setComparePrice(BigDecimal comparePrice) { this.comparePrice = comparePrice; }
    
    public String getCategoryId() { return categoryId; }
    public void setCategoryId(String categoryId) { this.categoryId = categoryId; }
    
    public String getBrand() { return brand; }
    public void setBrand(String brand) { this.brand = brand; }
    
    public List<String> getImages() { return images; }
    public void setImages(List<String> images) { this.images = images; }
    
    public List<String> getTags() { return tags; }
    public void setTags(List<String> tags) { this.tags = tags; }
    
    public BigDecimal getWeight() { return weight; }
    public void setWeight(BigDecimal weight) { this.weight = weight; }
    
    public Dimensions getDimensions() { return dimensions; }
    public void setDimensions(Dimensions dimensions) { this.dimensions = dimensions; }
    
    public Map<String, Object> getAttributes() { return attributes; }
    public void setAttributes(Map<String, Object> attributes) { this.attributes = attributes; }
    
    public List<Variant> getVariants() { return variants; }
    public void setVariants(List<Variant> variants) { this.variants = variants; }
    
    public Integer getStockQuantity() { return stockQuantity; }
    public void setStockQuantity(Integer stockQuantity) { this.stockQuantity = stockQuantity; }
    
    public Integer getLowStockThreshold() { return lowStockThreshold; }
    public void setLowStockThreshold(Integer lowStockThreshold) { this.lowStockThreshold = lowStockThreshold; }
    
    public Boolean getTrackInventory() { return trackInventory; }
    public void setTrackInventory(Boolean trackInventory) { this.trackInventory = trackInventory; }
    
    public Boolean getIsActive() { return isActive; }
    public void setIsActive(Boolean isActive) { this.isActive = isActive; }
    
    public Boolean getIsFeatured() { return isFeatured; }
    public void setIsFeatured(Boolean isFeatured) { this.isFeatured = isFeatured; }
    
    public BigDecimal getAverageRating() { return averageRating; }
    public void setAverageRating(BigDecimal averageRating) { this.averageRating = averageRating; }
    
    public Integer getReviewCount() { return reviewCount; }
    public void setReviewCount(Integer reviewCount) { this.reviewCount = reviewCount; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    
    // Helper methods
    public boolean isLowStock() {
        return trackInventory && stockQuantity <= lowStockThreshold;
    }
    
    public boolean isOutOfStock() {
        return trackInventory && stockQuantity <= 0;
    }
    
    public boolean hasDiscount() {
        return comparePrice != null && comparePrice.compareTo(price) > 0;
    }
    
    public BigDecimal getDiscountPercentage() {
        if (!hasDiscount()) return BigDecimal.ZERO;
        
        BigDecimal discount = comparePrice.subtract(price);
        return discount.divide(comparePrice, 4, BigDecimal.ROUND_HALF_UP)
                      .multiply(BigDecimal.valueOf(100));
    }
}

// Supporting classes
class Variant {
    private String id;
    private String name;
    private String sku;
    private BigDecimal price;
    private Integer stockQuantity;
    private Map<String, String> attributes; // color: "red", size: "L"
    private List<String> images;
    private Boolean isActive = true;
    
    // Constructors, getters, and setters
    public Variant() {}
    
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public String getSku() { return sku; }
    public void setSku(String sku) { this.sku = sku; }
    
    public BigDecimal getPrice() { return price; }
    public void setPrice(BigDecimal price) { this.price = price; }
    
    public Integer getStockQuantity() { return stockQuantity; }
    public void setStockQuantity(Integer stockQuantity) { this.stockQuantity = stockQuantity; }
    
    public Map<String, String> getAttributes() { return attributes; }
    public void setAttributes(Map<String, String> attributes) { this.attributes = attributes; }
    
    public List<String> getImages() { return images; }
    public void setImages(List<String> images) { this.images = images; }
    
    public Boolean getIsActive() { return isActive; }
    public void setIsActive(Boolean isActive) { this.isActive = isActive; }
}

File: services/admin-service/internal/models/product.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Product struct {
	ID            primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	Name          string             `bson:"name" json:"name" binding:"required,min=1,max=200"`
	Description   string             `bson:"description" json:"description"`
	Price         float64            `bson:"price" json:"price" binding:"required,min=0"`
	CategoryId    string             `bson:"categoryId" json:"categoryId" binding:"required"`
	Brand         string             `bson:"brand" json:"brand" binding:"required"`
	SKU           string             `bson:"sku" json:"sku"`
	StockQuantity int                `bson:"stockQuantity" json:"stockQuantity" binding:"min=0"`
	Images        []string           `bson:"images,omitempty" json:"images,omitempty"`
	Tags          []string           `bson:"tags,omitempty" json:"tags,omitempty"`
	IsActive      bool               `bson:"isActive" json:"isActive"`
	IsFeatured    bool               `bson:"isFeatured" json:"isFeatured"`
	Rating        float64            `bson:"rating" json:"rating"`
	ReviewCount   int                `bson:"reviewCount" json:"reviewCount"`
	SoldCount     int                `bson:"soldCount" json:"soldCount"`
	CreatedAt     time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt     time.Time          `bson:"updatedAt" json:"updatedAt"`
	CreatedBy     string             `bson:"createdBy,omitempty" json:"createdBy,omitempty"`
}

type CreateProductRequest struct {
	Name          string   `json:"name" binding:"required,min=1,max=200"`
	Description   string   `json:"description" binding:"required,min=10,max=2000"`
	Price         float64  `json:"price" binding:"required,min=0"`
	CategoryId    string   `json:"categoryId" binding:"required"`
	Brand         string   `json:"brand" binding:"required"`
	SKU           string   `json:"sku" binding:"required"`
	StockQuantity int      `json:"stockQuantity" binding:"min=0"`
	Images        []string `json:"images,omitempty"`
	Tags          []string `json:"tags,omitempty"`
	IsActive      *bool    `json:"isActive,omitempty"`
	IsFeatured    *bool    `json:"isFeatured,omitempty"`
}

type UpdateProductRequest struct {
	Name        string   `json:"name,omitempty" binding:"omitempty,min=1,max=200"`
	Description string   `json:"description,omitempty"`
	Price       *float64 `json:"price,omitempty" binding:"omitempty,min=0"`
	Category    string   `json:"category,omitempty"`
	Brand       string   `json:"brand,omitempty"`
	SKU         string   `json:"sku,omitempty"`
	Stock       *int     `json:"stock,omitempty" binding:"omitempty,min=0"`
	Images      []string `json:"images,omitempty"`
	Tags        []string `json:"tags,omitempty"`
	IsActive    *bool    `json:"is_active,omitempty"`
	IsFeatured  *bool    `json:"is_featured,omitempty"`
}
